src/main/java/jenkins/plugins/slack/SlackUserIdResolver.java

```diff
@@ -0,0 +1,167 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.plugins.slack;
+
+import hudson.model.AbstractBuild;
+import hudson.model.Run;
+import hudson.model.User;
+import hudson.scm.ChangeLogSet;
+import hudson.scm.ChangeLogSet.Entry;
+import hudson.tasks.MailAddressResolver;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import jenkins.plugins.slack.user.SlackUserProperty;
+import jenkins.scm.RunWithSCM;
+import org.apache.commons.lang.StringUtils;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpStatus;
+import org.apache.http.ParseException;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.util.EntityUtils;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class SlackUserIdResolver {
+
+    private static final Logger LOGGER = Logger.getLogger(SlackUserIdResolver.class.getName());
+
+    private static final String LOOKUP_BY_EMAIL_METHOD_URL_FORMAT = "https://slack.com/api/users.lookupByEmail?token=%s&email=%s";
+    private static final String SLACK_OK_FIELD = "ok";
+    private static final String SLACK_USER_FIELD = "user";
+    private static final String SLACK_ID_FIELD = "id";
+
+    private final String authToken;
+    private final CloseableHttpClient httpClient;
+    private final List<MailAddressResolver> mailAddressResolvers;
+
+    public SlackUserIdResolver(String authToken, CloseableHttpClient httpClient, List<MailAddressResolver> mailAddressResolvers) {
+        this.authToken = authToken;
+        this.httpClient = httpClient;
+        this.mailAddressResolvers = mailAddressResolvers != null ? mailAddressResolvers : MailAddressResolver.all();
+    }
+
+    public SlackUserIdResolver(String authToken, CloseableHttpClient httpClient) {
+        this(authToken, httpClient, null);
+    }
+
+    public List<String> resolveUserIdsForRun(Run run) {
```

Comment:
Can we suppress unchecked here? I don't thinks there's any way to work around it with the type signature of changesets?
```suggestion
    @SuppressWarnings("unchecked")
    public List<String> resolveUserIdsForRun(Run run) {
```